### DISCUSSION TITLE
How to get devices whose primary IP addresses belong to a certain prefix using the API?

### DISCUSSION THREAD

**Discussion by drygdryg** [netbox-community/netbox#11864] [Q&A] [answered]
How do I get a list of devices (dcim.device) whose primary IP addresses are in a certain prefix (ipam.prefix) using the REST API?
If this is not possible, is it possible to get a list of IP addresses that are in a certain prefix?
I can view IP addresses belonging to the prefix through web interface (e.g. https://demo.netbox.dev/ipam/prefixes/1/ip-addresses/), but I haven't found an API method for this.

**Comment by candlerb** [comment, marked as answer]
> How do I get a list of devices (dcim.device) whose primary IP addresses are in a certain prefix (ipam.prefix) using the REST API?

The available filters on `/api/dcim/devices/` are in the source code [here](https://github.com/netbox-community/netbox/blob/v3.4.5/netbox/dcim/filtersets.py#L793-L970).

The only query params which relate to primary ip are `has_primary_ip`, `primary_ip4_id` and `primary_ip6_id`.

However...

> If this is not possible, is it possible to get a list of IP addresses that are in a certain prefix?

Yes.  To find all the IP addresses within prefix 192.168.1.0/24, you would call `/api/ipam/ip-addresses/?parent=192.168.1.0%2F24` - code reference [here](https://github.com/netbox-community/netbox/blob/v3.4.5/netbox/ipam/filtersets.py#L479-L574).

Furthermore, you'll find each returned IP address has an `assigned_object` field, which will tell you the interface and device it belongs to (if any).  This solves your original problem, except you may get some IPs you're not interested in.

Adding `&assigned_to_interface=true` to the query parameters should filter out IPs not assigned to any interface, but you'll still get addresses which are not the *primary* address of a device, and addresses which are assigned to VM interfaces.

**Reply by drygdryg** [reply]
Thank you very much for the detailed answer!
